core/src/main/java/org/infinispan/persistence/util/PersistenceManagerCloseableSupplier.java

```diff
@@ -0,0 +1,159 @@
+   private final Executor executor;
+   private final PersistenceManager manager;
+   private final KeyFilter<K> filter;
+   private final InternalEntryFactory factory;
+   private final BlockingQueue<CacheEntry<K, V>> queue;
+   private final long timeout;
+   private final TimeUnit unit;
+
+   private final Lock closeLock = new ReentrantLock();
+   private final Condition closeCondition = closeLock.newCondition();
+
+   private boolean closed = false;
+   private AtomicReference<AdvancedCacheLoader.CacheLoaderTask<K, V>> taskRef = new AtomicReference<>();
+
+   public PersistenceManagerCloseableSupplier(Executor executor, PersistenceManager manager,
+                                              InternalEntryFactory factory, KeyFilter<K> filter, long timeout,
+                                              TimeUnit unit, int maxQueue) {
+      this.executor = executor;
+      this.manager = manager;
+      this.factory = factory;
+      this.filter = filter;
+      this.timeout = timeout;
+      this.unit = unit;
+      this.queue = new ArrayBlockingQueue<>(maxQueue);
+   }
+
+   class SupplierCacheLoaderTask implements AdvancedCacheLoader.CacheLoaderTask<K, V> {
+
+      @Override
+      public void processEntry(MarshalledEntry<K, V> marshalledEntry, AdvancedCacheLoader.TaskContext taskContext)
+              throws InterruptedException {
+         if (!taskContext.isStopped()) {
+            closeLock.lock();
+            try {
+               if (closed) {
+                  taskContext.stop();
+                  return;
+               }
+            } finally {
+               closeLock.unlock();
+            }
+            InternalCacheEntry<K, V> ice = PersistenceUtil.convert(marshalledEntry, factory);
+            // We do a read without acquiring lock
+            boolean stop = closed;
+            while (!stop) {
+               // If we were able to offer a value this means someone took from the queue so let us come back around main
+               // loop to offer all values we can
+               // TODO: do some sort of batching here - to reduce wakeup contention ?
+               if (queue.offer(ice, 100, TimeUnit.MILLISECONDS)) {
+                  closeLock.lock();
+                  try {
+                     // Wake up anyone waiting for a value
+                     closeCondition.signalAll();
+                  } finally {
+                     closeLock.unlock();
+                  }
+                  break;
+               }
+               // If we couldn't offer an entry check if we were completed concurrently
+               // We have to do in lock to ensure we see updated value properly
+               closeLock.lock();
+               try {
+                  stop = closed;
+               } finally {
+                  closeLock.unlock();
+               }
+            }
+         }
+      }
+   }
+
+   @Override
+   public CacheEntry<K, V> get() throws TimeoutException {
```

Comment:
Well, the other spliterator methods can be implemented quite trivially, if you don't need to support parallelism. I'd really like having spliterators "all the way down", instead of mixing iterators and suppliers, but I guess it's not that big of a difference.